errno/infoschema.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errno
+
+import (
+	"sync"
+	"time"
+)
+
+// The error summary is protected by a mutex for simplicity.
+// It is not expected to be hot unless there are concurrent workloads
+// that are generating high error/warning counts, in which case
+// the system probably has other issues already.
+
+// ErrorSummary summarizes errors and warnings
+type ErrorSummary struct {
+	ErrorCount   int
+	WarningCount int
+	FirstSeen    time.Time
+	LastSeen     time.Time
+}
+
+// instanceStatistics provide statistics for a tidb-server instance.
+type instanceStatistics struct {
+	sync.Mutex
+	global map[uint16]*ErrorSummary
+	users  map[string]map[uint16]*ErrorSummary
+	hosts  map[string]map[uint16]*ErrorSummary
+}
+
+var stats instanceStatistics
+
+func init() {
+	FlushStats()
+}
+
+// FlushStats resets errors and warnings across global/users/hosts
+func FlushStats() {
+	stats.Lock()
+	defer stats.Unlock()
+	stats.global = make(map[uint16]*ErrorSummary)
+	stats.users = make(map[string]map[uint16]*ErrorSummary)
+	stats.hosts = make(map[string]map[uint16]*ErrorSummary)
+}
+
+func copyMap(oldMap map[uint16]*ErrorSummary) map[uint16]*ErrorSummary {
+	newMap := make(map[uint16]*ErrorSummary, len(oldMap))
+	for k, v := range oldMap {
+		newMap[k] = v
```

Comment:
It's not a deep copy because `v` is a reference. This may lead to data race when accessing `ErrorSummary` concurrently.

Reply:
Thanks! I've fixed it and added a test to verify it.